scripts/component_hash_update/src/component_hash_update/download.py

```diff
@@ -0,0 +1,374 @@
+                for line in hashes.strip().split('\n')
+                if line.split('-')[-2] == "linux"
+                },
+             "nerdctl_archive": lambda hashes : {
+                line.split()[1].removesuffix('.tar.gz').split('-')[3] : line.split()[0]
+                for line in hashes.strip().split('\n')
+                if [x for x in line.split(' ') if x][1].split('-')[2] == "linux"
+                },
+            "runc": lambda hashes : {
+                parts[1].split('.')[1] : parts[0]
+                for parts in (line.split()
+                              for line in hashes.split('\n')[3:9])
+                },
+             "yq": lambda rhashes_bsd : {
+                 pair[0].split('_')[-1] : pair[1]
+                 # pair = (yq_<os>_<arch>, <hash>)
+                 for pair in ((line.split()[1][1:-1], line.split()[3])
+                     for line in rhashes_bsd.splitlines()
+                     if line.startswith("SHA256"))
+                 if pair[0].startswith("yq")
+                     and pair[0].split('_')[1] == "linux"
+                     and not pair[0].endswith(".tar.gz")
+                },
+            }
+
+    checksums_file = Path(subprocess.Popen(['git', 'rev-parse', '--show-toplevel'],
+                                           stdout=subprocess.PIPE).communicate()[0].rstrip().decode('utf-8')
+                     ) / CHECKSUMS_YML
+    logger.info("Opening checksums file %s...", checksums_file)
+    data, yaml = open_yaml(checksums_file)
+    s = requests.Session()
+
+    @cache
+    def _get_hash_by_arch(download: str, version: str) -> {str: str}:
+
+        hash_file = s.get(downloads[download]['url'].format(
+            version = version,
+            os = "linux",
+            ),
+                          allow_redirects=True)
+        hash_file.raise_for_status()
+        return download_hash_extract[download](hash_file.content.decode())
+
+
+    releases, tags = map(dict, partition(lambda r: r[1].get('tags', False), downloads.items()))
+    repos = {
+        'with_releases': [r['graphql_id'] for r in releases.values()],
+        'with_tags': [t['graphql_id'] for t in tags.values()],
+    }
+    response = s.post("https://api.github.com/graphql",
+                      json={'query': files(__package__).joinpath('list_releases.graphql').read_text(),
+                            'variables': repos},
+                      headers={
+                          "Authorization": f"Bearer {os.environ['API_KEY']}",
+                          }
+                      )
+    response.raise_for_status()
+    if 'x-ratelimit-used' in response.headers._store:
+        logger.info("Github graphQL API ratelimit status: used %s of %s. Next reset at %s",
+                    response.headers['X-RateLimit-Used'],
+                    response.headers['X-RateLimit-Limit'],
+                    datetime.fromtimestamp(int(response.headers["X-RateLimit-Reset"]))
+                    )
```

Comment:
I am wondering if this shouldn't be printed before the `raise_for_status` so that the person running the script has a clue if he his currently rate limited or not (assuming those headers are present on an error too), WDYT?

Also IIRC this rate limit doesn't apply for the other request in `get_hash` right? Which would means we have only request that counts towards those api rate limit? I it is maybe we could support running this script without even the API_KEY WDYT?

Reply:
Probably a good point, it looks like those header are indeed present: https://docs.github.com/en/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api#exceeding-the-rate-limit

Reply:
Yeah the get_hash does not use the API, it's only downloading assets. I'm not sure how those are rate-limited though ?
Also, not all of the assets are on github.com, some would be google infra, dl.k8s.io etc.

Running without API_KEY seems a bit hard, since we're using the graphql response as the source for which versions we are downloading. This would mean preserving the previous logic in parallel ? I'm not sure this is worth it.

Reply:
Potential improvement on the script would be to use the api for the release assets download (https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#get-a-release-asset) but that's another story (+ no idea how we'd deal with tags repo like gvisor)

Reply:
TIL about the rate limit around graphql I was more familiar with the github REST API ones, if it requires to save the result I don't think it is worth it too. I was expecting that you would be able to do this query without even being logged in, and repeat that a fair amount of time (5-10) but as this seems not true.

Reply:
Yeah the Graphql API is logged only, that's the downside.

Reply:
Yeah no worries feel free to resolve the thread it looks good to me as is, thanks!